haystack/components/builders/dynamic_chat_prompt_builder.py

```diff
@@ -0,0 +1,273 @@
+
+        :param runtime_variables: A list of template variable names you can use in chat prompt construction.
+        :type runtime_variables: Optional[List[str]]
+        """
+        runtime_variables = runtime_variables or []
+
+        # setup inputs
+        run_input_slots = {"prompt_source": List[ChatMessage], "template_variables": Optional[Dict[str, Any]]}
+        kwargs_input_slots = {var: Optional[Any] for var in runtime_variables}
+        component.set_input_types(self, **run_input_slots, **kwargs_input_slots)
+
+        # setup outputs
+        component.set_output_types(self, prompt=List[ChatMessage])
+        self.runtime_variables = runtime_variables
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+         Converts the `DynamicChatPromptBuilder` instance to a dictionary format, primarily for serialization purposes.
+
+        :return: A dictionary representation of the `DynamicChatPromptBuilder` instance, including its template
+        variables.
+        :rtype: Dict[str, Any]
+        """
+        return default_to_dict(self, runtime_variables=self.runtime_variables)
+
+    def run(self, prompt_source: List[ChatMessage], template_variables: Optional[Dict[str, Any]] = None, **kwargs):
+        """
+        Executes the dynamic prompt building process by processing a list of `ChatMessage` instances.
+        The last user message is treated as a template and rendered with the resolved pipeline variables and any
+        additional template variables provided.
+
+        You can provide additional template variables directly to this method, that are then merged with the variables
+        resolved from the pipeline runtime.
+
+        :param prompt_source: A list of `ChatMessage` instances. We make an assumption that the last user message has
+        the template for the chat prompt
+        :type prompt_source: List[ChatMessage]
+
+        :param template_variables: An optional dictionary of template variables. Template variables provided at
+        initialization are required to resolve pipeline variables, and these are additional variables users can
+        provide directly to this method.
+        :type template_variables: Optional[Dict[str, Any]]
+
+        :param kwargs: Additional keyword arguments, typically resolved from a pipeline, which are merged with the
+        provided template variables.
+
+        :return: A dictionary containing the key "prompt", which holds either the updated list of `ChatMessage`
+        instances or the rendered string template, forming the complete dynamic prompt.
+        :rtype: Dict[str, List[ChatMessage]
+        """
+        kwargs = kwargs or {}
+        template_variables = template_variables or {}
+        template_variables_combined = {**kwargs, **template_variables}
+        if not template_variables_combined:
+            raise ValueError(
+                "The DynamicChatPromptBuilder run method requires template variables, but none were provided. "
```

Comment:
Couldn't it be that a user doesn't have any parameters? Just sending the ChatMessage with no parameters could also happen, no? You probably have thought about that already, I am just curious.

Reply:
It could be, I'll make it a warning instead. Reading https://www.amazon.com/Philosophy-Software-Design-2nd/dp/173210221X/ changed my mind on this subject - exceptions should be thrown only in desperate situations.